libyul/backends/evm/StackHelpers.h

```diff
@@ -0,0 +1,376 @@
+/// abstracted shuffle operations.
+template</*ShuffleOperationConcept*/ typename ShuffleOperations>
+class Shuffler
+{
+public:
+	/// Executes the stack shuffling operations. Instantiates an instance of ShuffleOperations
+	/// in each iteration. Each iteration performs exactly one operation that modifies the stack.
+	/// After shuffle, source and target have the same size and all slots in the source layout are
+	/// compatible to the slots at the same target offset.
+	template<typename... Args>
+	static void shuffle(Args&&... args)
+	{
+		bool needsMoreShuffling = true;
+		// The shuffling algorithm should always terminate in polynomial time, but we provide a limit
+		// in case it does not terminate due to a bug.
+		size_t iterationCount = 0;
+		while (iterationCount < 1000 && (needsMoreShuffling = shuffleStep(std::forward<Args>(args)...)))
+			++iterationCount;
+		yulAssert(!needsMoreShuffling, "Could not create stack layout after 1000 iterations.");
+	}
+private:
+	// If dupping an ideal slot causes a slot that will still be required to become unreachable, then dup
+	// the latter slot first.
+	// @returns true, if it performed a dup.
+	static bool dupDeepSlotIfRequired(ShuffleOperations& _ops)
+	{
+		// Check if the stack is large enough for anything to potentially become unreachable.
+		if (_ops.sourceSize() < 15)
+			return false;
+		// Check whether any deep slot might still be needed later.
+		for (size_t sourceOffset: ranges::views::iota(0u, _ops.sourceSize() - 15))
+		{
+			// We need another copy of this slot.
+			if (_ops.sourceMultiplicity(sourceOffset) > 0)
+			{
+				// If this slot occurs again later, we skip this occurrence.
+				if (ranges::any_of(
+					ranges::views::iota(sourceOffset + 1, _ops.sourceSize()),
+					[&](size_t _offset) { return _ops.sourceIsSame(sourceOffset, _offset); }
+				))
+					continue;
+				// Bring up the target slot that would otherwise become unreachable.
+				for (size_t targetOffset: ranges::views::iota(0u, _ops.targetSize()))
+					if (!_ops.targetIsArbitrary(targetOffset) && _ops.isCompatible(sourceOffset, targetOffset))
+					{
+						_ops.pushOrDupTarget(targetOffset);
+						return true;
+					}
+			}
+		}
+		return false;
+	}
+	static void bringUpTargetSlot(ShuffleOperations& _ops, size_t _targetOffset)
+	{
+		std::list<size_t> toVisit{_targetOffset};
+		std::set<size_t> visited;
+
+		while (!toVisit.empty())
+		{
+			auto offset = *toVisit.begin();
+			toVisit.erase(toVisit.begin());
+			visited.emplace(offset);
+			if (_ops.targetMultiplicity(offset) > 0)
+			{
+				_ops.pushOrDupTarget(offset);
+				return;
+			}
+			// The desired target slot must already be somewhere else on stack right now.
+			for (auto nextOffset: ranges::views::iota(0u, std::min(_ops.sourceSize(), _ops.targetSize())))
+				if (
+					!_ops.isCompatible(nextOffset, nextOffset) &&
+					_ops.isCompatible(nextOffset, offset)
+				)
+					if (!visited.count(nextOffset))
+						toVisit.emplace_back(nextOffset);
```

Comment:
Can you explain the recursion here? Maybe as a property of the pre- and post-state of `bringUpTargetSlot`?

Reply:
Does the comment to the function I just pushed help :-)?

Reply:
Yep, I understand it now!